### ISSUE TITLE
UMD build with named and default export is strange

### ISSUE THREAD

**Issue by FredyC** [rollup/rollup#1044]
Here is [reproduction of the issue](http://rollupjs.org/?version=0.36.3&shareable=JTdCJTIyb3B0aW9ucyUyMiUzQSU3QiUyMmZvcm1hdCUyMiUzQSUyMnVtZCUyMiUyQyUyMm1vZHVsZU5hbWUlMjIlM0ElMjJteUJ1bmRsZSUyMiUyQyUyMmdsb2JhbHMlMjIlM0ElN0IlN0QlMkMlMjJtb2R1bGVJZCUyMiUzQSUyMiUyMiU3RCUyQyUyMm1vZHVsZXMlMjIlM0ElNUIlN0IlMjJuYW1lJTIyJTNBJTIybWFpbi5qcyUyMiUyQyUyMmNvZGUlMjIlM0ElMjJleHBvcnQlMjBmdW5jdGlvbiUyMGZvbygpJTIwJTdCJTVDbiU3RCU1Q24lNUNuZXhwb3J0JTIwZGVmYXVsdCUyMGZ1bmN0aW9uJTIwbWFpbigpJTIwJTdCJTVDbiU3RCU1Q24lMjIlN0QlNUQlN0Q=)

User gets the object like `{ default: function, foo: function }` so then to actually use the library it's necessary to use `lib.default()` to access main functionality. I understand why it works like this. However we have a use case that could actually benefit from exporting default function instead of a whole `exports` object.

See our [UMD build here](https://unpkg.com/stampit@3.0.3/dist/stampit.umd.js). We have a default export which is the main function. And this function gets also attached all other functions that are normally exported. By exporting this function, user would not lose any options, only gain in accessibility.

I am wondering if it could be configurable to have UMD bundle that points to default export?

**Comment by Rich-Harris** [comment]
See https://github.com/rollup/rollup/wiki/JavaScript-API#exports for an explanation of the `exports` option – by default, it will guess the mode based on what the entry module exports.

The only way you can avoid consumers having to do `lib.default()` is `exports: 'default'`. If you explicitly set that but also have named exports, it will throw, because that's logically incoherent – instead you would need to sacrifice the named exports and attach them to the default export as properties:

```js
assign(stampit, allUtilities);
export default stampit;
```

If you want to create an ES module build (e.g. so that Rollup and Webpack 2 users have a more seamless way of importing, and can tree-shake etc) then you'd just need to have a separate entry point – e.g. you could have an `index-umd.js` entry point that looks like the previous code block, and an `index-es.js` that looks like this:

```js
export default stampit;
export { methods, properties, ... };
```

Then, you'd create your `main` UMD bundle from `index-umd.js`, and your `module`/`jsnext:main` from `index-es.js`.

**Comment by FredyC** [comment]
Ok I see. So generally you are against adding it as built in optional feature to use default export in UMD build instead of `exports` object? I guess your approach will probably work as well, it's just not that nice :)

**Comment by jeanlescure** [comment]
Since I found this issue as one of the top google results when looking for a way to have both default and named exports work off of a single rollup output file, I'll be leaving my use case and final solution for anybody struggling with this situation.

My use case is simple:

I have a package called [short-unique-id](https://shortunique.id) which is fully Typescript, all source contained on a single `ts` file from which I generate two distribution files, one being the output of rollup (which coincidentally is the main entry point of the package) and a minified version which I generate by passing rollup's output file to terser.

I was trying to keep config to a minimum so I simply generated my dist file with rollup on CLI with the following command:

```sh
rollup -p @rollup/plugin-typescript -p @rollup/plugin-json -m -f umd -n ShortUniqueId --exports default -i src/index.ts -o dist/short-unique-id.js
```

This was initially ok because I **only had one export** on the `ts` file, but I also wanted to export a constant and an interface so that tsdoc would include them in the generated documentation.

When I exported those other members suddenly my named module `ShortUniqueId` became `default`.

To fix this now I have two source `ts` files.

`index.ts` the original file containing all the package's logic which I feed to `tsc` to generate only the `.d.ts` declaration file using the following command:

```
tsc -d --emitDeclarationOnly --rootDir src --outDir dist
```

and `dist.ts` which only contains the following:

```
import ShortUniqueId from './';

export default ShortUniqueId;
```

so now I pass `dist.ts` to rollup rather than `index.ts`:

```
rollup -p @rollup/plugin-typescript -p @rollup/plugin-json -m -f umd -n ShortUniqueId --exports default -i src/dist.ts -o dist/short-unique-id.js
```

Now `tsc` and `tsdoc` can use `index.ts` to generate the type definitions and docs, and `rollup` can use `dist.ts` to generate the named umd module.

I also added `default` as a static property to the class pointing to itself since webpack and rollup look for a default export on other packages that depend on my package if they use `allowSyntheticDefaultImports` in their tsconfig:

```js
export default class ShortUniqueId extends Function {
  static default: typeof ShortUniqueId = ShortUniqueId;

  //...
}
```

And that's it, now I have a single file that can be imported both as a named root module or as default.
